File: gr00t/Isaac-GR00T/examples/H20/robots/sdk_package/example/h10-w/motion/servo_joint.cpp
// get_cartesian_rotation_max_acc.cpp

#include <iostream>
#include <iomanip>
#include "h10-w/h10w_motion.h"
#include "common/common_def.h"
#include "common/h10w_def.h"
#include "h10-w/h10w_status.h"
#include "h10-w/h10w_system.h"
#include <algorithm>
#include <thread>

using namespace humanoid;
using namespace humanoid::h10w;
/// @brief
/// 以左臂为例从末端开始依次将各关节旋转0.3rad，再旋转回初始位置。初始位置应为0位，运行时需保持双臂不干涉且周围空旷。
int main() {
  std::cout << "=== realtime servo joint ===" << std::endl;
  H10wMotion motion;
  H10wStatus status;
  H10wSystem system;
  MoveMessageParams moveMsg;
  int result = status.getMoveMessage(moveMsg);
  if (result != 0) {
    std::cerr << "获取运动信息失败，任务终止" << std::endl;
    return -1;
  }
  bool leftArmNearZero =
      std::all_of(moveMsg.position.begin(), moveMsg.position.begin() + 7,
                  [](double value) { return value - 0.0 < 0.001; });
  if (!leftArmNearZero) {
    std::cerr << "为保证安全，请先将左臂回零" << std::endl;
    return -1;
  }
  result = system.powerControl(PowerState::POWER_ON);
  if (result != 0) {
    std::cerr << "上电失败" << std::endl;
    return -1;
  }
  result = system.clearError();
  if (result != 0) {
    std::cerr << "清错失败" << std::endl;
    return -1;
  }
  result = system.upperEnableControl(humanoid::EnableState::ENABLE_ON);
  if (result != 0) {
    std::cerr << "上半身使能失败" << std::endl;
    return -1;
  }

  motion.enableRealtimeCmd(true);
  RealtimeJointsParams joints;
  joints.left_arm_valid = false;
  joints.right_arm_valid = false;
  joints.torso_valid = false;
  const int joint_count =
      JointIndexValue::L_ARM_JOINT7 - JointIndexValue::L_ARM_JOINT1 + 1;
  for (int i = 0; i < joint_count; i++) {
    joints.left_arm[i] = moveMsg.position[i];
  }
  joints.time = 0.001;
  double step = 0.0001;
  bool servo_failed = false;
  for (int i = 0; i < joint_count; i++) {
    int time = 3000;
    while (time > 0) {
      joints.left_arm[i] += step;
      joints.left_arm_valid = true;
      if (motion.servoJoint(joints) != 0) {
        std::cout << "Servo joint failed." << std::endl;
        servo_failed = true;
        break;
      }
      std::this_thread::sleep_for(std::chrono::milliseconds(1));
      time--;
    }
    if (servo_failed) break;
    time = 0;
    while (time < 3000) {
      joints.left_arm[i] -= step;
      joints.left_arm_valid = true;
      if (motion.servoJoint(joints) != 0) {
        std::cout << "Servo joint failed." << std::endl;
        servo_failed = true;
        break;
      }
      std::this_thread::sleep_for(std::chrono::milliseconds(1));
      time++;
    }
    if (servo_failed) break;
  }
  motion.enableRealtimeCmd(false);
  std::vector<JointTarget> leftZeroPos{
      {static_cast<uint32_t>(JointIndexValue::L_ARM_JOINT1), 0},
      {static_cast<uint32_t>(JointIndexValue::L_ARM_JOINT2), 0},
      {static_cast<uint32_t>(JointIndexValue::L_ARM_JOINT3), 0},
      {static_cast<uint32_t>(JointIndexValue::L_ARM_JOINT4), 0},
      {static_cast<uint32_t>(JointIndexValue::L_ARM_JOINT5), 0},
      {static_cast<uint32_t>(JointIndexValue::L_ARM_JOINT6), 0},
      {static_cast<uint32_t>(JointIndexValue::L_ARM_JOINT7), 0}};
  result = motion.multiJointsMove(leftZeroPos, 0.1, 0.1);
  if (result != 0) {
    std::cout << "Move to zero failed." << std::endl;
  }

  result = motion.waitMove();
  if (result != 0) {
    std::cout << "Wait move to zero position failed." << std::endl;
    return -1;
  }
  return 0;
}

File: gr00t/Isaac-GR00T/examples/H20/robots/sdk_package/example_py/h10-w/motion/servo_joint.py
"""
以左臂为例从末端开始依次将各关节旋转0.3rad，再旋转回初始位置。
初始位置应为0位，运行时需保持双臂不干涉且周围空旷。
"""

import humanoid_sdk_py
from humanoid_sdk_py import h10w
import time
h10w_motion = h10w.H10wMotion()
h10w_status = h10w.H10wStatus()
h10w_system = h10w.H10wSystem()
time.sleep(1)
print("=== realtime servo joint ===")

ret, move_msg = h10w_status.getMoveMessage()
if ret != 0 or move_msg is None:
    print("获取运动信息失败，任务终止")
    exit(-1)

left_arm_near_zero = all(abs(value - 0.0) < 0.001 for value in move_msg.position[:7])
if not left_arm_near_zero:
    print("为保证安全，请先将左臂回零")
    exit(-1)

ret = h10w_system.powerControl(humanoid_sdk_py.PowerState.POWER_ON)
if ret != 0:
    print("上电失败")
    exit(-1)

ret = h10w_system.clearError()
if ret != 0:
    print("清除错误失败")
    exit(-1)

ret = h10w_system.upperEnableControl(humanoid_sdk_py.EnableState.ENABLE_ON)
if ret != 0:
    print("上半身使能失败")
    exit(-1)

ret = h10w_motion.enableRealtimeCmd(True)
if ret != 0:
    print("进入实时指令模式失败")
    exit(-1)


joints = h10w.RealtimeJointsParams()
joints.left_arm_valid = False
joints.right_arm_valid = False
joints.torso_valid = False
joints.time = 0.001
for i in range(7):
    joints.left_arm[i] = move_msg.position[i]

print("move message: current pos")
print(joints.left_arm)

step = 0.0001
servo_failed = False
for i in range(7):
    time_count = 3000
    while time_count > 0:
        joints.left_arm[i] += step
        joints.left_arm_valid = True
        ret = h10w_motion.servoJoint(joints)
        if ret != 0:
            print("Servo joint failed.")
            servo_failed = True
            break
        time.sleep(0.001)
        time_count -= 1
    if servo_failed:
        break

    time_count = 0
    while time_count < 3000:
        joints.left_arm[i] -= step
        joints.left_arm_valid = True
        ret = h10w_motion.servoJoint(joints)
        if ret != 0:
            print("Servo joint failed.")
            servo_failed = True
            break
        time.sleep(0.001)
        time_count += 1
    if servo_failed:
        break

ret = h10w_motion.enableRealtimeCmd(False)
if ret != 0:
    print("关闭实时指令失败")
    exit(-1)

left_homing_pos = [
    (h10w.L_ARM_JOINT1, 0.0),
    (h10w.L_ARM_JOINT2, 0.0),
    (h10w.L_ARM_JOINT3, 0.0),
    (h10w.L_ARM_JOINT4, 0.0),
    (h10w.L_ARM_JOINT5, 0.0),
    (h10w.L_ARM_JOINT6, 0.0),
    (h10w.L_ARM_JOINT7, 0.0),
]

ret = h10w_motion.multiJointsMove(left_homing_pos, 0.1, 0.1)
if ret != 0:
    print("Move to zero failed")
    exit(-1)

ret = h10w_motion.waitMove()
if ret != 0:
    print("Move to zero timeout")
    exit(-1)
